src/HIndent/ModulePreprocessing/CommentRelocation.hs

```diff
@@ -0,0 +1,336 @@
+{-# LANGUAGE FlexibleContexts    #-}
+{-# LANGUAGE GADTs               #-}
+{-# LANGUAGE RankNTypes          #-}
+{-# LANGUAGE RecordWildCards     #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TypeApplications    #-}
+
+-- | Comment relocation for pretty-printing comments correctly.
+--
+-- HIndent gathers all comments above a function, an import, a module
+-- declaration, etc. For example, HIndent formats the following code
+--
+-- > f :: Int
+-- > f = 1
+-- >
+-- > -- A comment between f and g
+-- >
+-- > -- Another comment between f and g
+-- >
+-- > g :: Int
+-- > g = 2
+--
+-- to
+--
+-- > f :: Int
+-- > f = 1
+-- >
+-- > -- A comment between f and g
+-- > -- Another comment between f and g
+-- > g :: Int
+-- > g = 2
```

Comment:
It would be nice if comment separation would be maintained, but maybe we can push this to a separate issue.

Reply:
Yes. We need to change this behavior. I'll try it later. The issue is here: https://github.com/mihaimaruseac/hindent/issues/608.